pymongo/mongo_client.py

```diff
@@ -2334,6 +2338,8 @@ def run(self) -> T:
                             raise
                         self._retrying = True
                         self._last_error = exc
+                        if self._client.topology_description.topology_type == TOPOLOGY_TYPE.Sharded:
```

Comment:
I think we need to clear the deprioritized servers after all retry attempts (pass or fail).

Reply:
Good catch, thanks!

Reply:
I would also expect multiple failures to append to the deprioritized server list. What does the spec say?

Reply:
The spec seems to suggest that only a single retry will occur, but using a list of deprioritized servers suggests to me that multiple failures would result in multiple servers being deprioritized.

Reply:
Yes, I meant after the retry completes (whether it succeeds or not).

Reply:
_ClientConnectionRetryable is a temp object so I don't see a reason to clear the list.

Reply:
Ah, very nice!